frame/uniques/src/lib.rs

```diff
@@ -0,0 +1,1182 @@
+			let origin = ensure_signed(origin)?;
+
+			let mut details = Asset::<T, I>::get(&class, &instance)
+				.ok_or(Error::<T, I>::Unknown)?;
+			let class_details = Class::<T, I>::get(&class).ok_or(Error::<T, I>::Unknown)?;
+			ensure!(class_details.freezer == origin, Error::<T, I>::NoPermission);
+
+			details.is_frozen = true;
+			Asset::<T, I>::insert(&class, &instance, &details);
+
+			Self::deposit_event(Event::<T, I>::Frozen(class, instance));
+			Ok(())
+		}
+
+		/// Re-allow unprivileged transfer of an asset instance.
+		///
+		/// Origin must be Signed and the sender should be the Freezer of the asset `class`.
+		///
+		/// - `class`: The class of the asset to be thawed.
+		/// - `instance`: The instance of the asset to be thawed.
+		///
+		/// Emits `Thawed`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::thaw())]
+		pub(super) fn thaw(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId,
+			#[pallet::compact] instance: T::InstanceId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			let mut details = Asset::<T, I>::get(&class, &instance)
+				.ok_or(Error::<T, I>::Unknown)?;
+			let class_details = Class::<T, I>::get(&class).ok_or(Error::<T, I>::Unknown)?;
+			ensure!(class_details.admin == origin, Error::<T, I>::NoPermission);
+
+			details.is_frozen = false;
+			Asset::<T, I>::insert(&class, &instance, &details);
+
+			Self::deposit_event(Event::<T, I>::Frozen(class, instance));
+			Ok(())
+		}
+
+		/// Disallow further unprivileged transfers for a whole asset class.
+		///
+		/// Origin must be Signed and the sender should be the Freezer of the asset `class`.
+		///
+		/// - `class`: The asset class to be frozen.
+		///
+		/// Emits `ClassFrozen`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::freeze_class())]
+		pub(super) fn freeze_class(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			Class::<T, I>::try_mutate(class, |maybe_details| {
+				let details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+				ensure!(&origin == &details.freezer, Error::<T, I>::NoPermission);
+
+				details.is_frozen = true;
+
+				Self::deposit_event(Event::<T, I>::ClassFrozen(class));
+				Ok(())
+			})
+		}
+
+		/// Re-allow unprivileged transfers for a whole asset class.
+		///
+		/// Origin must be Signed and the sender should be the Admin of the asset `class`.
+		///
+		/// - `class`: The class to be thawed.
+		///
+		/// Emits `ClassThawed`.
+		///
+		/// Weight: `O(1)`
+		#[pallet::weight(T::WeightInfo::thaw_class())]
+		pub(super) fn thaw_class(
+			origin: OriginFor<T>,
+			#[pallet::compact] class: T::ClassId
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+
+			Class::<T, I>::try_mutate(class, |maybe_details| {
+				let details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+				ensure!(&origin == &details.admin, Error::<T, I>::NoPermission);
```

Comment:
I made a comment in the `assets` pallet, and I will make it again here.

I think it is strange that the `admin` is the one who can thaw an asset. I would have expected either a single role for freezing and thawing, or a specific role for the two actions.

You cannot, with this implementation, have only the ability to freeze / thaw. To thaw, you must have an admin, who has access to change other more sensitive parts of the asset.

A single role for both actions makes it so that you can choose to opt in or out of the whole feature. Otherwise, setting two roles allows you to separate who controls the ability to freeze and thaw, but not overload the ability to thaw with other implications about your asset.

Reply:
The point of making them different is that freezing is a much less problematic operating than thawing and any other admin functions. Freezing just (temporarily) prevents movement. Its effects can not lead to unrecoverable consequences. Thawing, minting and burning are all fundamentally more sensitive. So I do not think that they belong in a single class; in operational systems, one might imagine freezing to be a 1-of-N among many employees or an hot wallet sitting in a bot. Thawing and the other operations would be something more inaccessible and of less chance for compromise.

We could reasonably split out a key for thawing, if there's clear need among users.